File: python-playwright/framework/products/parabank/steps.py
"""Parabank business steps (Layer 2) — the bank workflows a test cares about."""

from __future__ import annotations

from playwright.sync_api import Page

from framework import config
from framework.core import get_logger

from .data import Customer
from .pages import AccountsOverviewPage, OpenAccountPage, RegisterPage, TransferFundsPage


class ParabankSteps:
    def __init__(self, page: Page) -> None:
        self.page = page
        self.base = config.PARABANK_URL
        # Parabank can be slow; give actions room.
        self.page.set_default_timeout(45_000)
        self.register_page = RegisterPage(page)
        self.overview = AccountsOverviewPage(page)
        self.open_account_page = OpenAccountPage(page)
        self.transfer_page = TransferFundsPage(page)
        self.log = get_logger("ParabankSteps")

    def register(self, customer: Customer) -> str:
        self.log.info("register user %s", customer.username)
        self.register_page.open(self.base)
        return self.register_page.register(customer)

    def open_savings_account(self) -> str:
        self.log.info("open a savings account")
        new_id = self.open_account_page.open(self.base).open_account(OpenAccountPage.TYPE_SAVINGS)
        self.log.info("opened account %s", new_id)
        return new_id

    def transfer(self, amount: int, to_account: str, from_account: str | None = None) -> str:
        self.log.info("transfer %s -> %s", amount, to_account)
        return self.transfer_page.open(self.base).transfer(amount, to_account, from_account)

    def overview_account_ids(self) -> list[str]:
        return self.overview.open(self.base).account_ids()

File: ts-playwright/src/products/parabank/steps.ts
// Parabank business steps (Layer 2) — the bank workflows a test cares about.
import { Page } from '@playwright/test';
import { URLS } from '../../config';
import { getLogger, Logger } from '../../core/logger';
import { Customer } from './data';
import {
  AccountsOverviewPage,
  OpenAccountPage,
  RegisterPage,
  TransferFundsPage,
} from './pages';

export class ParabankSteps {
  private base = URLS.parabank;
  private registerPage: RegisterPage;
  private overview: AccountsOverviewPage;
  private openAccountPage: OpenAccountPage;
  private transferPage: TransferFundsPage;
  private log: Logger;

  constructor(private page: Page) {
    this.page.setDefaultTimeout(45_000); // Parabank can be slow
    this.registerPage = new RegisterPage(page);
    this.overview = new AccountsOverviewPage(page);
    this.openAccountPage = new OpenAccountPage(page);
    this.transferPage = new TransferFundsPage(page);
    this.log = getLogger('ParabankSteps');
  }

  async register(customer: Customer): Promise<string> {
    this.log.info(`register user ${customer.username}`);
    await this.registerPage.open(this.base);
    return this.registerPage.register(customer);
  }

  async openSavingsAccount(): Promise<string> {
    this.log.info('open a savings account');
    await this.openAccountPage.open(this.base);
    const id = await this.openAccountPage.openAccount(OpenAccountPage.TYPE_SAVINGS);
    this.log.info(`opened account ${id}`);
    return id;
  }

  async transfer(amount: number, toAccount: string, fromAccount?: string): Promise<string> {
    this.log.info(`transfer ${amount} -> ${toAccount}`);
    await this.transferPage.open(this.base);
    return this.transferPage.transfer(amount, toAccount, fromAccount);
  }

  async overviewAccountIds(): Promise<string[]> {
    await this.overview.open(this.base);
    return this.overview.accountIds();
  }
}
